ports/esp32/machine_adc.c

```diff
@@ -31,178 +32,414 @@
 
 #include "driver/gpio.h"
 #include "driver/adc.h"
+#include "esp_adc_cal.h"
 
 #include "py/runtime.h"
 #include "py/mphal.h"
 #include "modmachine.h"
 
+
+#define DEFAULT_VREF 1100
+
+
+typedef struct _madcblock_obj_t {
+    mp_obj_base_t base;
+    adc_unit_t unit_id;
+    mp_int_t bits;
+    adc_bits_width_t width;
+    esp_adc_cal_characteristics_t *characteristics[ADC_ATTEN_MAX];
+} madcblock_obj_t;
+
+STATIC madcblock_obj_t madcblock_obj[] = {
+    #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32C3
+    {{&machine_adcblock_type}, ADC_UNIT_1, 12, -1, {0}},
+    {{&machine_adcblock_type}, ADC_UNIT_2, 12, -1, {0}},
+    #elif CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
+    {{&machine_adcblock_type}, ADC_UNIT_1, 13, -1, {0}},
+    {{&machine_adcblock_type}, ADC_UNIT_2, 13, -1, {0}},
+    #endif
+};
+
+#define ADCBLOCK1 (&madcblock_obj[0])
+#define ADCBLOCK2 (&madcblock_obj[1])
+
+
 typedef struct _madc_obj_t {
     mp_obj_base_t base;
+    madcblock_obj_t *block;
+    adc_channel_t channel_id;
     gpio_num_t gpio_id;
-    adc1_channel_t adc1_id;
+    adc_atten_t atten;
 } madc_obj_t;
 
-STATIC const madc_obj_t madc_obj[] = {
+STATIC madc_obj_t madc_obj[] = {
```

Comment:
This really should stay as `const` otherwise it eats up valuable RAM.

Is the `atten` value the only thing that needs to change?  If so then it can go in a separate non-const array, possibly of a single byte per channel.

Reply:
Yes, it's just the `atten` field, so I can move this out too.

Reply:
Done (I think…)